core/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -465,129 +473,166 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
+                new ParseField(Fields.HIGHLIGHT));
+        parser.declareObject((map, value) -> {
+            Map<String, SearchHitField> fieldMap = get(Fields.FIELDS, map, new HashMap<String, SearchHitField>());
+            fieldMap.putAll(value);
+            map.put(Fields.FIELDS, fieldMap);
+        }, SearchHit::parseFields, new ParseField(Fields.FIELDS));
+        parser.declareObject((map, value) -> map.put(Fields._EXPLANATION, value), SearchHit::parseExplanation,
+                new ParseField(Fields._EXPLANATION));
+        parser.declareObject((map, value) -> map.put(NestedIdentity._NESTED, value), NestedIdentity::fromXContent,
+                new ParseField(NestedIdentity._NESTED));
+        parser.declareObject((map, value) -> map.put(Fields.INNER_HITS, value), SearchHit::parseInnerHits,
+                new ParseField(Fields.INNER_HITS));
+        parser.declareStringArray((map, list) -> map.put(Fields.MATCHED_QUERIES, list), new ParseField(Fields.MATCHED_QUERIES));
+        parser.declareField((map, list) -> map.put(Fields.SORT, list), SearchSortValues::fromXContent, new ParseField(Fields.SORT),
+                ValueType.OBJECT_ARRAY);
+    }
+
+    public static SearchHit createFromMap(Map<String, Object> values) {
+        String id = get(Fields._ID, values, null);
+        String type = get(Fields._TYPE, values, null);
+        NestedIdentity nestedIdentity = get(NestedIdentity._NESTED, values, null);
+        Map<String, SearchHitField> fields = get(Fields.FIELDS, values, null);
+
+        SearchHit searchHit = new SearchHit(-1, id, new Text(type), nestedIdentity, fields);
+        searchHit.index = get(Fields._INDEX, values, null);
+        searchHit.score(get(Fields._SCORE, values, DEFAULT_SCORE));
+        searchHit.version(get(Fields._VERSION, values, -1L));
+        searchHit.sortValues(get(Fields.SORT, values, SearchSortValues.EMPTY));
+        searchHit.highlightFields(get(Fields.HIGHLIGHT, values, null));
+        searchHit.sourceRef(get(SourceFieldMapper.NAME, values, null));
+        searchHit.explanation(get(Fields._EXPLANATION, values, null));
+        searchHit.setInnerHits(get(Fields.INNER_HITS, values, null));
+        List<String> matchedQueries = get(Fields.MATCHED_QUERIES, values, null);
+        if (matchedQueries != null) {
             searchHit.matchedQueries(matchedQueries.toArray(new String[matchedQueries.size()]));
         }
+        ShardId shardId = get(Fields._SHARD, values, null);
+        String nodeId = get(Fields._NODE, values, null);
         if (shardId != null && nodeId != null) {
             searchHit.shard(new SearchShardTarget(nodeId, shardId));
         }
         searchHit.fields(fields);
         return searchHit;
     }
 
-    private static Explanation parseExplanation(XContentParser parser) throws IOException {
+    private static <T> T get(String key, Map<String, Object> map, T defaultValue) {
```

Comment:
Can you at least change it to:

    ```
@SuppressWarnings("unchecked")
    private static <T> T get(String key, Map<String, Object> map, T defaultValue) {
        return (T) map.getOrDefault(key, defaultValue);
    }
```

I don't want us to reimplement core stuff just to avoid an explicit cast